Add Matter client example for webOS OSE

cabin15 opened the pull request:
#### Problem
* There is no support of taking advantage of Matter on webos.

#### Change overview
- Adding cross compile support of Matter on webOS OSE along with client example.
- Include Matter client example for webOS OSE (So far, only device discovery API works on webOS OSE. Still implementing the feature)
- Introduce webos as a individual platform layer. (Original source is forked from src/platform/Linux to src/platform/webos)
- BLE and Wi-Fi support will be done later by using the webos platform API.

#### Testing
How was this tested? (at least one bullet point required)
* If new unit tests are not added, why not?
Unit tests are not support as of now and will be covered as part of next release update

* If new integration tests are not added, why not?
Integration and CI build support will be added as part of next update

* If manually tested, what platforms controller and device platforms were manually tested, and how?
chip-tool is cross-compiled for webOS OSE and tested.
chip-tool works as expected regarding discovery, pairing, commissioning window operation,read/write attribute.

Build of chip-tool for webOS OSE can be done by executing below script.
$ . scripts/examples/webos_example.sh

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=13292) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=13292) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=13292) it.</sub>

cabin15 commented:
Dear Reviewers,

This is Sungmok Shin from LG Electronics Korea.

This pull request is about introducing Matter build and application testing environment on opensource platform called "webOS OSE".

You can check more about webOS OSE on below url.
- https://www.webosose.org/

Currently, we have port Matter on webOS OSE, and implement the simple background service that provides device discovery functionality via IPC mechanism called "luna-service2" used on webOS OSE.

Our objective is to introduce more API on this service for webOS application developer to use this API to implement their own Matter IoT application on webOS OSE.


I understand that there are still lots of things that needs to be fixed for this PR to be merged due to lack of my short experiences.
Think of this PR as a rough start for our contribution.

I would like you guys to review this at any point. Any reviews are welcomed.

Thank you
Sungmok Shin

mspang commented on build/config/compiler/BUILD.gn:
```diff
@@ -204,6 +204,13 @@ config("strict_warnings") {
     "-Wvla",
   ]
 
+  if (target_os == "webos") {
```
This needs to say current_os not target_os.

The way that target_os is being used in this PR is not correct. It's just setting the default toolchain, and the initial value for current_os in the default toolchain. They have the same value space. We shouldn't have target_os = "webos" set current_os = "linux".

cabin15 replied:
Thank you for your review, Michael.
Let me fix this by using 'current_os == "webos"'.

cabin15 replied:
Dear Michael,
Your comment is fixed on below commit.
https://github.com/project-chip/connectedhomeip/pull/13292/commits/332aeb0e597be919476b0ac4350c53b72d096726

I fixed the build script to have the initial value of current_os as 'webos'.

Now, branching the current_os == "webos" works on build/config/compiler/BUILD.gn
Let me know if this is the right fix.
Thank you.

mspang reviewed (dismissed):
current_os / target_os confusion needs to be resolved

cabin15 commented on src/crypto/crypto.gni:
```diff
@@ -24,7 +24,7 @@ if (chip_crypto == "") {
       current_os == "zephyr" || current_os == "mbed") {
     chip_crypto = "mbedtls"
   } else {
-    chip_crypto = "openssl"
```
Sorry that I missed branching this.
Let me fix this.

cabin15 replied:
Dear Jonathan,
I modified the crypto.gni to use 'mbedtls' when current_os is webos.
Let me know if this does not introduce any side effect.
Thank you.

andy31415 commented on examples/chip-tool-webos/README.md:
```diff
@@ -0,0 +1,143 @@
+
+## Build Instructions
+
+Step 1: Build and Flash the webOS OSE
```
For all other matter platforms we generally provide the ability to compile from a docker image, including allowing the use of `chip-build-vscode` image to compile any build (including arm64 cross compile).

Is this something that could be done for webOS? as a developer I would find it quite complex to go through all the steps below to just validate that my changes keep compiling on webOS.

cabin15 replied:
Dear Andrei,
I appreciate your kind feedback.
Let me check your opinion and get back to you.
Thank you.

cabin15 replied:
Dear Andrei,
To follow your guide, I made a script (scripts/example/webos_example.sh) to build chip-tool for webos using webos ndk. In this way, developer who wants to implement their application on webOS OSE could easily work on it. I will modify the README soon. Until then, could you plz review the rest of the changes? Thank you in advance.

cabin15 commented:
Dear Reviewers,

Could someone could shine on my problem regarding processing this PR?

I have a 2 questions as follows.

1) 22 workflows awaiting approval
It seems like there are testing job "Waiting for status to be reported" such as Ameba, Build Android, Build Darwin etc.
Is this job started when time passes? or someone with authorization should trigger this?
Last time, notification mail has been sent to me that those jobs are failed due to my PR. So, I fixed it and added additional commits on this PR. So, I wanna see if my PR still break the existing build or not.


2) pullapprove added "review - pending" label 12 ago.
Does this mean that peer review is not on progress? (ex) reviewer is not able to review this until further action.)
Or this is because of "Unresolved conversations" not resolved yet?
I wish I could knew if peer review is still on going.

Thank you
Sungmok Shin

bzbarsky-apple commented on examples/chip-tool-webos/commands/clusters/ModelCommand.h:
```diff
@@ -0,0 +1,61 @@
+#include <app/chip-zcl-zpro-codec.h>
+#include <lib/core/CHIPEncoding.h>
+
+class ModelCommand : public CHIPCommand
```
This looks like a copy/paste of examples/chip-tool, without changes to most of the code, right?  If the problem is that chip-tool in its current form does not compile/run on webos (why not?), then the right fix is to fix that, not copy all the code.  Especially because this copy/paste is already out of date; chip-tool has changed under it.

cabin15 replied:
@bzbarsky-apple 
First of all, I appreciate your kind review.

webos has no problem compiling the examples/chip-tool.
We just intended to add webos specific changes on chip-tool. That's why I created separate example folder for webos.

But I think your review is right.
In this case, chip-tool resources in the examples/chip-tool-webos will be out-of-date which will introduce build or runtime error in the future.

Referencing other platform, it seems like they set their build to use the latest chip-tool located on examples/chip-tool.
I think webos should follows the same.

Let me check your review and will get back to you once fix is ready.
Thank you.

cabin15 replied:
To keep the chip-tool compiled on webOS as up-to-date, I decided to remove the examples/chip-tool-webos.
From now on, examples/chip-tool will be built as the example for webOS OSE.
Thank you.

cabin15 commented:
Dear Reviewers,
Currently, workflow on this PR it not run but skipped all the time.
I suspect that I might not able to run workflow due to below rules.

"GitHub Actions: Maintainers must approve first time contributor workflow runs"

Could someone please approve the workflow run for me?
Thank you.

cabin15 commented:
@bzbarsky-apple 
I appreciate your help running the workflow.

cabin15 commented:
@mspang @bzbarsky-apple @jmeg-sfy @andy31415 
Could you plz review this?
Thank you in advance.

andy31415 commented:
> @mspang @bzbarsky-apple @jmeg-sfy @andy31415 Could you plz review this? Thank you in advance.

We are  under one week from feature freeze and being asked to review a PR with 11K delta lines ....

andy31415 commented on scripts/examples/webos_example.sh:
```diff
@@ -0,0 +1,53 @@
+
+# Download webOS OSE NDK
+echo "##### Download webOS OSE NDK #####"
+wget http://10.178.94.187/ndk/opt/webos-sdk-x86_64-cortexa7t2hf-neon-vfpv4-toolchain-2.14.1.g.host.manifest
```
what is this IP address? 

SDK should be available for everyone, generally in a sdk image.

Also using `build_examples.py` is preffered to shell scripts.

cabin15 replied:
@andy31415 
I appreciate your review.
I will check the way to change this as a public url for everyone.
Also, I'll check to see if I can use build_examples.py for webos example build.
Always thank you.

cabin15 replied:
@andy31415 
I had changed the private IP address to public url (sourceforge).

I tried to use git-lfs, but failed due to file size limited to 2.2G.
Fortunately, sourceforge doesn't have file size limit.

I still have a problem about time (about 50~70 minutes) spent for downloading 2.2G.
I think I should find a faster hosting ASAP.

Thank you for your review.

andy31415 reviewed (changes requested):
build example seems to use a private IP address to download SDKs.

Please create a webos docker image and it should be based on generally public SDKs (usually github or some official company provided addresses without NDA).

andy31415 commented on build/config/webos/pkg_config.gni:
```diff
@@ -0,0 +1,134 @@
+# Copyright (c) 2013 The Chromium Authors. All rights reserved.
+# Copyright (c) 2020 Project CHIP Authors
```
what are the deltas from `build/config/linux/pkg_config.gni` ? can we reuse instead of using a copy?

cabin15 replied:
@andy31415 
As you suggested, I deleted below files and re-use the one from linux.
- build/config/webos/pkg_config.gni
- build/config/webos/pkg-config.py

Thank you for your review.

andy31415 commented on src/platform/webos/CHIPLinuxStorage.cpp:
```diff
@@ -0,0 +1,321 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Is this a copy of linux? Name says "CHIPLinuxStorage" but platform is webos.

cabin15 replied:
@andy31415 
As I'm not ready for using the own storage implementation for webOS, let me remove this and re-use the one from Linux.
Thank you for your review.

andy31415 commented on src/platform/webos/DnssdImpl.h:
```diff
@@ -0,0 +1,173 @@
+/*
```
Can we re-use linux-based impl files and headers instead of copying?

cabin15 replied:
@andy31415 
Yes. Let me remove this and re-use the one from Linux.
Thank you for your review.

andy31415 commented on src/platform/webos/KeyValueStoreManagerImpl.h:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ *    @file
+ *          Platform-specific implementation of KVS for linux.
```
this is in webos, so likely should not say "linux". Also should reuse not copy&paste

cabin15 replied:
@andy31415 
As you suggested, let me remove this and re-use the one from Linux.
Thank you for your review.

cabin15 commented:
> > @mspang @bzbarsky-apple @jmeg-sfy @andy31415 Could you plz review this? Thank you in advance.
> 
> We are under one week from feature freeze and being asked to review a PR with 11K delta lines ....
I'm sorry for asking you guys to review a PR with hugh delta.

Actually, I think problem for reviewing might be the resources under src/platform.
Rest of the build script change might be simple to review.

For your information, as you already saw, all the stuffs under src/platform/webos is copy&paste from src/platform/Linux.
It is because we are working on platform adaptation layer for webOS by replicating and modifying the src/platform/Linux to src/platform/webos.

Those changes are not done yet. So, I created first PR with just copying the src/platform/Linux to src/platform/webos.
You can think of this commit as the initial commit for webOS Port.

Let me first fix your reviewed point.
Also let me try reusing the existing resource and remove the unnecessary stuff under src/platform/webos.

If it is not too much trouble, could you plz suggest how we can minimize the reviewing effort of src/platform/webos?
Thank you.

woody-apple commented:
Approving, and fast tracking, given > 3 days for review, and these are mostly platform changes. @cabin15 please follow up with some CI to verify that this platform builds?

cabin15 commented:
> Approving, and fast tracking, given > 3 days for review, and these are mostly platform changes. @cabin15 please follow up with some CI to verify that this platform builds?

Dear Justin,

I really appreciate your approval and fast tracking.

Yes, I'm planning to add CI for webos platform build verification as well as other platform.
(Also, as @andy31415 suggested, I should come up with webos docker image things).
If you don't mind, can I work for CI in the next PR since this PR is getting bigger and bigger?

Thank you
Sungmok Shin

andy31415 commented:
> > Approving, and fast tracking, given > 3 days for review, and these are mostly platform changes. @cabin15 please follow up with some CI to verify that this platform builds?
> 
> Dear Justin,
> 
> I really appreciate your approval and fast tracking.
> 
> Yes, I'm planning to add CI for webos platform build verification as well as other platform. (Also, as @andy31415 suggested, I should come up with webos docker image things). If you don't mind, can I work for CI in the next PR? I'm afraid that this PR is getting bigger and bigger which give reviewers a hard time checking too many deltas.
> 
> Thank you Sungmok Shin

I stand by my blocking of the PR as long as it uses a private IP address to download a script and execute.
please create a separate SDK docker image and have appropriate commands - I imagine webos should have some public way to download SDKs (or if it does not  have one, one should be created).

We have a requirement in matter that officially supported builds (i.e. things in the main tree) are available for use by all developers, not behind an NDA and stable (hence docker images).

cabin15 commented:
> > > Approving, and fast tracking, given > 3 days for review, and these are mostly platform changes. @cabin15 please follow up with some CI to verify that this platform builds?
> > 
> > 
> > Dear Justin,
> > I really appreciate your approval and fast tracking.
> > Yes, I'm planning to add CI for webos platform build verification as well as other platform. (Also, as @andy31415 suggested, I should come up with webos docker image things). If you don't mind, can I work for CI in the next PR? I'm afraid that this PR is getting bigger and bigger which give reviewers a hard time checking too many deltas.
> > Thank you Sungmok Shin
> 
> I stand by my blocking of the PR as long as it uses a private IP address to download a script and execute. please create a separate SDK docker image and have appropriate commands - I imagine webos should have some public way to download SDKs (or if it does not have one, one should be created).
> 
> We have a requirement in matter that officially supported builds (i.e. things in the main tree) are available for use by all developers, not behind an NDA and stable (hence docker images).

@andy31415 
Your review is always appreciated.

I replaced private IP address with public url where everyone can download webOS NDK.
It has been verified that script works as expected on public network as well.

Now the remaining work seems like,
1) Integrate webos_example.sh to build_examples.py
2) Prepare webOS docker image for testing
3) Cleaning up the webos platform folder


If it is not too much trouble, I wonder if I can work on above items once this PR could be merged.

Like you mentioned eariler, this PR has too many deltas which waste reviewers valuable times.

As most of those deltas are on webos platform folder which does not break the CI, I wonder if this PR could be approved for now so that remaining work could continue confortably.

Me and my colleague is working on BLE and Wi-Fi adaptation for webos platform layer.
Sooner or later, there will be major changes on src/platform/webos which includes cleaning up this folder.

If you think those 1)~3) items are blocker, let me know. I would be glad to fix those with best effort.

Thank you.
Sungmok Shin

cabin15 commented:
@mspang 
Could you plz check if your review is fixed in the latest?
Thank you.

woody-apple commented:
It’s fine to follow up with CI, we’ve done the same for previous platforms

cabin15 commented:
> It’s fine to follow up with CI, we’ve done the same for previous platforms

I appreciate your understanding. Let me follow up with CI in the next PR.

woody-apple commented:
Note still needs changes requested to be resolved, or we need sufficient time for @andy31415 and @mspang to respond.

cabin15 commented:
> Note still needs changes requested to be resolved, or we need sufficient time for @andy31415 and @mspang to respond.

@woody-apple 
Thank you for your comment.
I think I should fix the "changes requested" first by any means.
Let me get back to here once I resolve the changes requested.
Thank you.

woody-apple commented:
> > Note still needs changes requested to be resolved, or we need sufficient time for @andy31415 and @mspang to respond.
> 
> @woody-apple Thank you for your comment. I think I should fix the "changes requested" first by any means. Let me get back to here once I resolve the changes requested. Thank you.

Thank you for your contributions!

cabin15 commented:
@andy31415 
I pushed the "clean-up" commit which resolved your review.
As you suggested, I removed the copy&pasted file and re-use the one from Linux.
Other than the file you pointed, I also remove the platform resources which could be re-used on Linux.
Also, on BLE and Connectivity related files, I removed the whole body and added "NOT IMPLEMENTED" return for future implementation.
There are still many deltas, but much more smaller.

Could you plz review this PR again?
Thank you.

cabin15 commented:
Hope zap failure in zap generation and restyling has nothing to do with this PR.

cabin15 commented:
Could anyone trigger CI job for this PR? since I'm not authorized to do it.
I just wanna make sure that CI is not broken. :)

andy31415 commented on scripts/examples/webos_example.sh:
```diff
@@ -0,0 +1,53 @@
+
+# Install webOS OSE NDK
+echo "##### Install webOS OSE NDK #####"
+sudo ./webos-sdk-x86_64-cortexa7t2hf-neon-vfpv4-toolchain-2.14.1.g.sh
```
why is sudo needed here?

I do not trust scripts to "downloads something off the internet and executes as root".

andy31415 replied:
Looking at the script, it seems to support `-d` for target_sdk_dir. Lets use that and not root.

andy31415 replied:
We should also check and download only if the sdk does not exist. The SDK itself seems very large: 2.1GB.

Downloading every compilation would be extreamly slow.

jmeg-sfy replied:
+100 @andy31415 -> Usage of a public git repo is more preferable than a wget scripting

jmeg-sfy replied:
> why is sudo needed here?
> 
> I do not trust scripts to "download something off the internet and execute as root".

Yes this is totally sketchy !

jmeg-sfy replied:
If you guys really want to go this way then you will have to verify a legit signature of the package... (SHA, MD5) 
Once again - @cabin15, it would be better to use the common scheme of external repositories being located into the third_party folder as a git submodule

cabin15 replied:
@andy31415, @jmeg-sfy 
Thank you for your review.
According to your review, I stopped using 'wget'. And I registered the webOS SDK resources as a submodule located under third_party.
Would this be fine for the baseline start?
Thank you.

andy31415 commented on scripts/examples/webos_example.sh:
```diff
@@ -0,0 +1,53 @@
+echo "##### Building by ninja #####"
+ninja -C out/host
+
+# Remove webOS OSE NDK archive
```
should removal of downloaded files be done after the installation?

cabin15 replied:
@andy31415 
Thank you for your review.
I removed this part for developer to this artifact for their own usage.
Would this be fine?

jmeg-sfy commented on src/platform/webos/BLEManagerImpl.cpp:
```diff
@@ -0,0 +1,207 @@
+bool BLEManagerImpl::SendWriteRequest(BLE_CONNECTION_OBJECT conId, const Ble::ChipBleUUID * svcId, const Ble::ChipBleUUID * charId,
+                                      chip::System::PacketBufferHandle pBuf)
+{
+    return true;
```
These stubs will be full filled later on other PRs right ?

cabin15 replied:
@jmeg-sfy 
Sure. Yes. We are currently working for this and all stubs will be filled out on next PR.
Thank you for your review.

cabin15 commented:
@andy31415 @jmeg-sfy 
I appreciate your reviews.
Let me go through each every one and fix those.
Thank you.

jmeg-sfy commented:
> @andy31415 @jmeg-sfy I appreciate your reviews. Let me go through each every one and fix those. Thank you.

Please ping me once done

jmeg-sfy commented on .gitmodules:
```diff
@@ -184,3 +184,6 @@
 	url = https://github.com/ATmobica/mcuboot.git
+[submodule "third_party/webos_sdk"]
+	path = third_party/webos_sdk
+	url = https://github.com/cabin15/webos-ose-ndk
```
You use your own account @cabin15  in the github link path , no one has a more official repo (i guess for LG)
that fine but you know i would not recommend that for long term solution

cabin15 replied:
@jmeg-sfy 
You are right.
Our team is about on the start line about matter work and contribution.
It would be so kind if you could think this PR as a start baseline.
I'll drive internal discussion for providing webOS SDK as LG official repo.
Thank you for your review.

jmeg-sfy commented on scripts/examples/webos_example.sh:
```diff
@@ -0,0 +1,42 @@
+
+# Install webOS OSE NDK
+echo "##### Install webOS OSE NDK #####"
+third_party/webos_sdk/v2.14.1/webos-sdk-x86_64-cortexa7t2hf-neon-vfpv4-toolchain-2.14.1.g.sh -d third_party/webos_sdk/v2.14.1 -y
```
Sources are not directly clear in plain in your git , that fine i would suggest , here you pull an artifact and extract it

cabin15 replied:
@jmeg-sfy 
Yes. I tried to put all resources as a plain file, but failed due to its size.
I think I should have internal discussion with relevant team to reduce SDK size to be fit so that all resources could be stored as plain form in next PR.
Thank you for your review and understanding.

andy31415 commented on src/platform/webos/Logging.cpp:
```diff
@@ -0,0 +1,52 @@
+/* See Project CHIP LICENSE file for licensing information. */
```
This license blurb is different fro the rest

cabin15 replied:
@andy31415 
I modified the license blurb of src/platform/webos/Logging.cpp just like other files.
Thank you for your review.

cabin15 commented:
@andy31415 
Thank you for dismissing the stale review.
As you mentioned, it is already fixed.

cabin15 commented:
@andy31415 
Thank you for triggering the CI.

Build Android failed due to "No space left on device" which seems not relevant with this PR.
I think failure on Line Code Base should be fixed.
Let me double check this.
Thank you.

cabin15 commented:
Failure of Lint check is fixed on:
- https://github.com/project-chip/connectedhomeip/pull/13292/commits/fcd9b88ed623a8fe7bab0e854353d829aedf739a

cabin15 commented:
Conflict on .gitmodules (due to the add of perfecto) is resolved.

cabin15 commented:
It would so kind if someone could run the CI. :)

cabin15 commented:
Regarding 3 failed CI, I'm trying to see if this PR has anything to do with this.

1. Build Android - 
INFO    /opt/android/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: final link failed: No space left on device
--> It seems not relevant with this PR. But I have idea why Build Android succeeded on other PR.

2. Test Suites Darwin
2022-02-08 17:17:11.907 ERROR   17:17:11.544 - TEST OUT  : [1644340631543] [9491:57923] CHIP: [TOO]  ***** Test Failure: currentLevel value mismatch: expected 128 but got 114
--> It seems like the failure during run time testing.

3. Build Darwin.
Failed during "Run Framework Tests".
--> Seems like not relevant.

Could someone shine on me with this?
Is this just the problem on the CI of other platform? or this PR has anything to do with this?

Thank you in advance.

cabin15 commented:
Let me fetch the latest upstream to see if broken CI succeeds.

cabin15 commented:
@andy31415 @woody-apple @jmeg-sfy ,
I started to wonder if thia PR intorduces the Android CI failure by adding webos sdk as submodule which exceeds the space for Android CI.
https://github.com/project-chip/connectedhomeip/runs/5126795951?check_suite_focus=true

Would this be a blocker for merging this?
If so, could you plz guide me how to fix this?
Forgive me since I'm new to CI things.

cabin15 commented:
@andy31415 @woody-apple @jmeg-sfy 
Android CI had failed since webos_sdk is added on submodule list.

Android CI failure log is at:
- https://github.com/project-chip/connectedhomeip/runs/5126795951

{code}
    INFO    [938/940] ld -shared lib/jni/arm64-v8a/libCHIPController.so
    INFO    FAILED: lib/jni/arm64-v8a/libCHIPController.so lib/jni/arm64-v8a/libCHIPController.map
    INFO    /opt/android/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang++ -Wl,--gc-sections -march=armv8-a -O0 -fPIC -Werror -Wl,--fatal-warnings -Wl,-z,defs -fdiagnostics-color -Wl,-Map,lib/jni/arm64-v8a/libCHIPController.map -Wl,--gc-sections @lib/jni/arm64-v8a/libCHIPController.so.rsp -o lib/jni/arm64-v8a/libCHIPController.so -shared
    INFO    /opt/android/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: final link failed: No space left on device
    INFO    clang++: error: linker command failed with exit code 1 (use -v to see invocation)
{code}

It keeps displaying "No space left on device" which I believe because of additional submodule added by this PR.

For testing purpose, I removed webos_sdk on the submodule, and Android CI succeeded.
    
It seems like docker size limit for Android CI introduces this.
At this point, I have no clear solution to fix this error with maintaining webos_sdk on submodule.
So, for now, I removed webos_sdk on the submodule and let scripts/examples/webos_example.sh to check out the webos sdk repo when building webos application.

It would be so kind if anyone trigger the CI again.

Also, could you plz let me know if above change would be fine?
If so, it would be so happy if you guys could go over this PR for merging.
Always I appreciate your help.

jmeg-sfy commented:
> @andy31415 @woody-apple @jmeg-sfy Android CI had failed since webos_sdk is added on submodule list.
> 
> Android CI failure log is at:
> 
> * https://github.com/project-chip/connectedhomeip/runs/5126795951
> 
> {code} INFO [938/940] ld -shared lib/jni/arm64-v8a/libCHIPController.so INFO FAILED: lib/jni/arm64-v8a/libCHIPController.so lib/jni/arm64-v8a/libCHIPController.map INFO /opt/android/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang++ -Wl,--gc-sections -march=armv8-a -O0 -fPIC -Werror -Wl,--fatal-warnings -Wl,-z,defs -fdiagnostics-color -Wl,-Map,lib/jni/arm64-v8a/libCHIPController.map -Wl,--gc-sections @lib/jni/arm64-v8a/libCHIPController.so.rsp -o lib/jni/arm64-v8a/libCHIPController.so -shared INFO /opt/android/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: final link failed: No space left on device INFO clang++: error: linker command failed with exit code 1 (use -v to see invocation) {code}
> 
> It keeps displaying "No space left on device" which I believe because of additional submodule added by this PR.
> 
> For testing purpose, I removed webos_sdk on the submodule, and Android CI succeeded.
> 
> It seems like docker size limit for Android CI introduces this. At this point, I have no clear solution to fix this error with maintaining webos_sdk on submodule. So, for now, I removed webos_sdk on the submodule and let scripts/examples/webos_example.sh to check out the webos sdk repo when building webos application.
> 
> It would be so kind if anyone trigger the CI again.
> 
> Also, could you plz let me know if above change would be fine? If so, it would be so happy if you guys could go over this PR for merging. Always I appreciate your help.

Woody and Andrei are the persons taking care of the CI.  This would be nice to understand why there is No space left on device indeed.
May be your archive contains many unused stuffs . i can see your are pulling toolchains for neon ...  what about the one already available on the SDK , those might work !? ... does a LG tv compile on a specific toolchains only.. ?

cabin15 commented:
> > Also, could you plz let me know if above change would be fine? If so, it would be so happy if you guys could go over this PR for merging. Always I appreciate your help.
> 
> Woody and Andrei are the persons taking care of the CI. This would be nice to understand why there is No space left on device indeed. May be your archive contains many unused stuffs . i can see your are pulling toolchains for neon ... what about the one already available on the SDK , those might work !? ... does a LG tv compile on a specific toolchains only.. ?

@jmeg-sfy 
Thank you for your opinion.
As a matter of fact, I'm already investigating this by running the android CI locally to reproduce this.
Like you said, this would be a good opportunity yo undetstand how CI works on this repo.
Let me investigate this and get back to here.

mspang commented on build/toolchain/webos/BUILD.gn:
```diff
@@ -0,0 +1,25 @@
+
+import("webos_toolchain.gni")
+
+print(target_os)
```
Please remove this print.

cabin15 replied:
print is removed.
Thank you for your review.

mspang commented on build/toolchain/webos/BUILD.gn:
```diff
@@ -0,0 +1,25 @@
+
+webos_toolchain("webos") {
+  toolchain_args = {
+    target_os = "linux"
```
"webos"

cabin15 replied:
@mspang 
My bad. Fixed as "webos"
Thank you for your review.

mspang commented on build/toolchain/webos/webos_toolchain.gni:
```diff
@@ -0,0 +1,37 @@
+  print("WebOS_ToolChian")
+  gcc_toolchain(target_name) {
+    ar = ar_webos
+    cc = cc_webos + "  -O2 -Wno-format-security -Wno-error -Wno-cpp"
```
Please don't specify warning or optimization flags in the toolchain as these are controlled by args.

Please don't build with -Wno-error

cabin15 replied:
Let me remove this and control further actions by args.
Thank you for your review.

mspang commented on build/toolchain/webos/webos_toolchain.gni:
```diff
@@ -0,0 +1,37 @@
+import("//build/toolchain/gcc_toolchain.gni")
+
+template("webos_toolchain") {
+  print("WebOS_ToolChian")
```
Please remove this print.

cabin15 replied:
print is removed.
Thank you for your review.

mspang commented on src/platform/webos/BLEManagerImpl.h:
```diff
@@ -0,0 +1,252 @@
+
+struct BLEAdvConfig
+{
+    char * mpBleName;
```
= nullptr ?

cabin15 replied:
Let me initialize the mpBleName as nullptr.
Thank you for your review.

mspang commented on src/platform/webos/BLEManagerImpl.h:
```diff
@@ -0,0 +1,252 @@
+struct BLEAdvConfig
+{
+    char * mpBleName;
+    uint32_t mAdapterId;
```
= 0?

and more below

cabin15 replied:
Let me initialize the int variables as 0.
Thank you for your review.

mspang commented on src/platform/webos/BUILD.gn:
```diff
@@ -0,0 +1,132 @@
+
+static_library("webos") {
+  if (enable_syslog) {
+    print("### enable syslog logging ###")
```
Please remove this print.

cabin15 replied:
print is removed.
Thank you for your review.

mspang commented on src/platform/webos/BUILD.gn:
```diff
@@ -0,0 +1,132 @@
+    ]
+  }
+
+  print("### enable_wifi = ${chip_enable_wifi}")
```
Please remove this print. You can run "gn args --list" to see all the args and their value.

cabin15 replied:
print is removed.
Thank you for your review.

mspang commented on src/platform/webos/BUILD.gn:
```diff
@@ -0,0 +1,132 @@
+    #public_deps += [ "dbus/wpa" ]
+  }
+
+  print("### enable_ble = ${chip_enable_ble}")
```
Please remove this print.

cabin15 replied:
print is removed.
Thank you for your review.

cabin15 commented:
@mspang 
Thank you for your review.
I will fix those and get back.
BTW, it would be so helpful if you could approve and run the ci for this PR so that I could recognize the current status.

cabin15 commented:
@mspang 
Your reviews are resolved in the latest commit.
https://github.com/project-chip/connectedhomeip/pull/13292/commits/4b21399d7e02262266a658765e9e31e768a5ba4a

Could you plz continue reviewing this?
I appreciate your review effort.

cabin15 commented:
@andy31415 @woody-apple @jmeg-sfy @mspang 
Gentle reminder for continuing review this PR.
Thank you in advance.

cabin15 commented:
Latest conflict has been resolved.

cabin15 commented:
@andy31415 @woody-apple @jmeg-sfy @mspang 
It would be so kind if you let me know what I have to do next to get this PR merged.
Should I wait for extra approval

cabin15 commented:
It would be so kind if anyone could trigger the CI for me.

cabin15 commented:
@jmeg-sfy 
Thank you for triggering the CI. 👍 

@andy31415 @woody-apple @mspang @jmeg-sfy 
It would be so kind if you could consider merging this PR, so that I could move on.

cabin15 commented:
@andy31415 
Gentle reminder for checking this PR for next step.
Thank you.

andy31415 commented:
/rebase

cabin15 commented:
> /rebase

@andy31415 
Rebase is done. Thank you.
